### DISCUSSION TITLE
[RFC FS-1108] Undentation frenzy

### DISCUSSION THREAD

**Discussion by Happypig375** [fsharp/fslang-design#602] [Language and core library RFC discussions]
RFC: https://github.com/fsharp/fslang-design/blob/main/preview/FS-1108-undentation-frenzy.md
Suggestions:

- [Allow undentation for constructors](https://github.com/fsharp/fslang-suggestions/issues/724)
- [Relax indentation rules on mutually recursive functions when using with CE](https://github.com/fsharp/fslang-suggestions/issues/868)
- [Relax some of the indentation rules](https://github.com/fsharp/fslang-suggestions/issues/470)
- [Add new undentation rule for match with](https://github.com/fsharp/fslang-suggestions/issues/833)
- [Relax indentation rules on anonymous records](https://github.com/fsharp/fslang-suggestions/issues/786)
- [Allow undentation in method and constructor calls](https://github.com/fsharp/fslang-suggestions/issues/931)
- [Alignment requirements for mutually recursive functions are non-intuitive](https://github.com/fsharp/fslang-suggestions/issues/601)
- [Mutual Recursion with Async Workflow Possible Compiler Issue](https://github.com/dotnet/fsharp/issues/3326)
- [Composed function leads to unexpected indentation warning](https://github.com/dotnet/fsharp/issues/10852)
- [Warning only when function application is inside SynExpr.IfThenElse](https://github.com/dotnet/fsharp/issues/10929)

Pull request: https://github.com/dotnet/fsharp/pull/11772

**Comment by charlesroddie** [comment]
This RFC made the consistency adjustment:

```fsharp
type A(x:string) = class end

let x = [ // This was OK before this RFC and seemed a reasonable way to format code
    "data"
    ]

let y = A( // adjusted by this RFC for consistency, also seems reasonable
    "data"
    )
```

There is a side effect that the following code is allowed:

```
let x' =
    [ // this bracket begins on a line with only whitespace before; can we be strict in this case and not permit subsequent undentations
    "data" // this line should be required to be indented further
    ]

let y' = // similarly
    A(
    "data"
    )
```

The rule could be that undentations in lines following unclosed brackets/constructors are allowed when the opening bracket is of the form `let _ = [opening bracket/constructor]`.

Thoughts @Happypig375 @dsyme ?

**Reply by dsyme** [reply]
It is a little odd that is accepted, though it was also accepted in F# 5.0 it seems

**Reply by dsyme** [reply]
That is, the first (`x'`) was accepted without warning in F# 5.0, the second wasn't.

**Reply by charlesroddie** [reply]
Does this require a new language suggestion to fix?
